### PR TITLE
Add referrerpolicy to list of relevant mutations

### LINKED ISSUE #26388: Treat referrerpolicy attribute as a "relevant mutation" for image elements
Spec change: https://github.com/whatwg/html/pull/5434
Tests: https://github.com/web-platform-tests/wpt/pull/23309
Code: https://github.com/servo/servo/blob/64dee1fa116870eb76c3b8b54090afce9c4d1d13/components/script/dom/htmlimageelement.rs#L1592-L1599

We should hopefully see new test results if we run `./mach test-wpt tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/relevant-mutations.html` after adding the attribute to the list of attributes that cause the image data to be updated.

### PR DESCRIPTION
**Pull request by PeterZhizhin**
`img.referrerPolicy` change now mutates the image.
Not all `referrerPolicy` attribute changes result in an image update event.
Only valid changes are reflected.

All referrerpolicy tests inside `relevant-mutations.html` WPT test now pass.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #26388

- [x] These have tests:
```
./mach test-wpt tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/relevant-mutations.html
```

### CODE REVIEW

**Comment by PeterZhizhin** [comment]
I don't fully understand the test results. However, it appears that there are some relevant tests with unexpected results:
```
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLImageElement interface: attribute referrerPolicy
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLImageElement interface: document.createElement("img") must inherit property "referrerPolicy" with the proper type
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLImageElement interface: new Image() must inherit property "referrerPolicy" with the proper type
  ▶ Unexpected subtest result in /custom-elements/reactions/HTMLImageElement.html:
  └ PASS [expected FAIL] referrerPolicy on HTMLImageElement must enqueue an attributeChanged reaction when adding a new attribute
  ▶ Unexpected subtest result in /custom-elements/reactions/HTMLImageElement.html:
  └ PASS [expected FAIL] referrerPolicy on HTMLImageElement must enqueue an attributeChanged reaction when replacing an existing attribute
```

However, I cannot run this test locally. I tried running:
```
.\mach test-wpt .\tests\wpt\web-platform-tests\html\dom\idlharness.https.html
```
This is due to #26431 (it runs the test in headless mode).

I am not sure what to do next. Should I edit `servo\tests\wpt\metadata\html\dom\idlharness.https.html.ini` and remove the expected `HTMLImageElement` failures?

What about the other tests? Is this me or they are just flaky?

**Comment by jdm** [comment]
Oh, only the results in the public/filtered-wpt-errorsummary.log artifact matter; the others are filtered into public/intermittents.log because they are tests with known intermittent failures. And yes, removing those expected failures from the ini file is the correct choice here.

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /html/dom/idlharness.https.html?include=HTML.*:
  └ PASS [expected FAIL] HTMLImageElement interface: attribute referrerPolicy
```
One missing one.

**Comment by PeterZhizhin** [comment]
> 
> 
> ```
>   ▶ Unexpected subtest result in /html/dom/idlharness.https.html?include=HTML.*:
>   └ PASS [expected FAIL] HTMLImageElement interface: attribute referrerPolicy
> ```
> 
> One missing one.

Sorry, added removed this expectation as well.

File: components/script/dom/htmlimageelement.rs [hunk 1]
```diff
@@ -1542,6 +1557,31 @@ impl HTMLImageElementMethods for HTMLImageElement {
         }
     }
 
+    // https://html.spec.whatwg.org/multipage/#attr-img-referrerpolicy
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs:1560]: This link should be https://html.spec.whatwg.org/multipage/#dom-img-referrerpolicy.

File: components/script/dom/htmlimageelement.rs [hunk 2]
```diff
@@ -1542,6 +1557,31 @@ impl HTMLImageElementMethods for HTMLImageElement {
...
+    fn GetReferrerPolicy(&self) -> Option<DOMString> {
+        let element = self.upcast::<Element>();
+        let current_policy_value = element.get_string_attribute(&local_name!("referrerpolicy"));
+        Some(get_correct_referrerpolicy_from_raw_token(
+            &current_policy_value,
+        ))
+    }
+
+    // https://html.spec.whatwg.org/multipage/#attr-img-referrerpolicy
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs:1569]: https://html.spec.whatwg.org/multipage/#dom-img-referrerpolicy

File: components/script/dom/webidls/HTMLImageElement.webidl [hunk 3]
```diff
@@ -27,6 +27,8 @@ interface HTMLImageElement : HTMLElement {
   readonly attribute unsigned long naturalHeight;
   readonly attribute boolean complete;
   readonly attribute USVString currentSrc;
+  [CEReactions]
+           attribute DOMString? referrerPolicy;
```
Comment (jdm) [inline components/script/dom/webidls/HTMLImageElement.webidl:31]: The spec does not include a ?.

File: components/script/dom/htmlimageelement.rs [hunk 4]
```diff
@@ -1420,6 +1420,21 @@ pub fn parse_a_sizes_attribute(value: DOMString) -> SourceSizeList {
     SourceSizeList::parse(&context, &mut parser)
 }
 
+pub(self) fn get_correct_referrerpolicy_from_raw_token(token: &DOMString) -> DOMString {
```
Comment (jdm) [inline components/script/dom/htmlimageelement.rs:1423]: What's the purpose of `pub(self)`?

**Review by jdm — CHANGES_REQUESTED** [review]
Looks good! Please squash your changes into a single commit, too.

File: components/script/dom/htmlimageelement.rs (hunk 4 above)
Comment (PeterZhizhin) [reply components/script/dom/htmlimageelement.rs]: I didn't compile without this line before. Thank you for helping me with this!

**Comment by PeterZhizhin** [comment]
I've addressed all review comments. Also squashed all commits into one.

**Comment by jdm** [comment]
@bors-servo r+
Thanks!

**Pull request merged**